File: sliding-window-maximum/sliding-window-maximum.java
class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
        List<Integer> ans = new ArrayList<>();
        Deque<Integer> dq = new ArrayDeque<>();
        for(int i=0; i<k; i++){
            while(!dq.isEmpty() && nums[i]>=nums[dq.peekLast()]) dq.pollLast();
            dq.offerLast(i);
        }
        ans.add(nums[dq.peekFirst()]);
        for(int i=k; i<nums.length; i++){
            if(dq.peekFirst()==i-k) dq.pollFirst();
            while(!dq.isEmpty() && nums[i]>=nums[dq.peekLast()]) dq.pollLast();
            dq.offerLast(i);
            ans.add(nums[dq.peekFirst()]);
        }
        return ans.stream().mapToInt(i->i).toArray();
    }
}

// Sliding maximum

File: sliding-window-maximum/sliding-window-maximum.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        vector<int> ans;
        deque<int> dq;
        for(int i=0; i<k; i++){
            while(!dq.empty() && nums[i]>=nums[dq.back()]) dq.pop_back();
            dq.push_back(i);
        }
        ans.push_back(nums[dq.front()]);
        for(int i=k; i<nums.size(); i++){
            if(dq.front() == i-k) dq.pop_front();
            while(!dq.empty() && nums[i]>=nums[dq.back()]) dq.pop_back();
            dq.push_back(i);
            ans.push_back(nums[dq.front()]);
        }
        return ans;
    }
};

// Sliding Window Maximum 